app/uk/gov/hmrc/gform/gform/SummaryExprSubstituter.scala

```diff
@@ -122,43 +160,80 @@ object SummarySubstituter {
-  ) {
-
-    /** This method is used to replace the Sum expression with Adds
-      * It also substitutes the hidden FormCtx in the sumExpr with Constant(0)
-      */
-    def replaceSumWithAdds(sumExpr: Expr): Expr = {
-      val sumExprFcIds = sumExpr.allFormComponentIds()
-      val indexedSumFcIdMap = exprMap.collect {
-        case (FormCtx(fcId), evalResult)
-            if sumExprFcIds
-              .map(_.baseComponentId)
-              .contains(fcId.baseComponentId) && fcId.modelComponentId.indexedComponentId.isIndexed =>
-          fcId -> evalResult
-      }.toMap
-
-      val indexedSumExprFcIds = indexedSumFcIdMap.map(_._1)
-      val indices = indexedSumExprFcIds
-        .map(fcId => fcId.modelComponentId.indexedComponentId)
-        .collect { case IndexedComponentId.Indexed(_, index) =>
-          index
-        }
-        .toList
-        .sorted
-        .distinct
-
-      import FormComponentIdSubstituter._
-      val substitutions = new FormComponentIdSubstitutions()
-      val baseSumExpr: Expr =
-        implicitly[Substituter[FormComponentIdSubstitutions, Expr]].substitute(substitutions, sumExpr)
-      val fcs = indexedSumExprFcIds.map(_.modelComponentId.removeIndex.toFormComponentId).toList.distinct
-      indices
-        .map(i => ExprUpdater(baseSumExpr, i, fcs))
-        .foldLeft[Expr](Constant("0")) { case (acc, e) =>
-          Add(acc, e)
-        }
+}
+
+case class SummarySubstitutions(
+  exprMap: Map[Expr, ExpressionResult],
+  data: VariadicFormData[SourceOrigin.OutOfDate]
+) {
+
+  def replaceSumWithAdds(sumExpr: Expr): Expr = {
+    val atlHelper = new ATLNestedAddHelper(exprMap, sumExpr)
+    val groupHelper = new GroupNestedAddHelper(data, sumExpr)
+    if (atlHelper.indexedFormComponentIds.nonEmpty) {
+      atlHelper.toNestedAdd()
+    } else if (groupHelper.indexedFormComponentIds.nonEmpty) {
+      groupHelper.toNestedAdd()
+    } else {
+      sumExpr
     }
   }
 }
+
+abstract class NestedAddHelper(sumExpr: Expr) {
+
+  val sumExprFcIds = sumExpr.allFormComponentIds()
+
+  def computeIndexedFormComponentIds(): List[FormComponentId]
+  def computeIndices(): List[Int]
+
+  lazy val indexedFormComponentIds: List[FormComponentId] = computeIndexedFormComponentIds()
+  lazy val indices: List[Int] = computeIndices()
+
+  def toNestedAdd(): Expr = {
+    import FormComponentIdSubstituter._
+    val substitutions = new FormComponentIdSubstitutions()
+    val baseSumExpr: Expr =
+      implicitly[Substituter[FormComponentIdSubstitutions, Expr]].substitute(substitutions, sumExpr)
+    val fcs = indexedFormComponentIds.map(_.modelComponentId.removeIndex.toFormComponentId).distinct
+    indices
+      .map(i => ExprUpdater(baseSumExpr, i, fcs))
+      .foldLeft[Expr](Constant("0")) { case (acc, e) =>
+        Add(acc, e)
+      }
+  }
+
+  protected def extractIndices(fcIds: List[FormComponentId]): List[Int] =
+    fcIds.flatMap(_.modelComponentId.maybeIndex).sorted.distinct
```

Comment:
Calling `distinct` before `sorted` to avoid unnecessary sorting of duplicate elements.